pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/utils/RealTimeRegexpMatcher.java

```diff
@@ -0,0 +1,158 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.Automaton;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.CharacterRunAutomaton;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.RegExp;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.State;
+import org.apache.pinot.segment.local.utils.nativefst.automaton.Transition;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableArc;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableFST;
+import org.apache.pinot.segment.local.utils.nativefst.mutablefst.MutableState;
+import org.roaringbitmap.IntConsumer;
+
+public class RealTimeRegexpMatcher {
+  private final String _regexQuery;
+  private final MutableFST _fst;
+  private final Automaton _automaton;
+  private final IntConsumer _dest;
+
+  public RealTimeRegexpMatcher(String regexQuery, MutableFST fst, IntConsumer dest) {
+    _regexQuery = regexQuery;
+    _fst = fst;
+    _dest = dest;
+
+    _automaton = new RegExp(_regexQuery).toAutomaton();
+  }
+
+  public static void regexMatch(String regexQuery, MutableFST fst, IntConsumer dest) {
+    RealTimeRegexpMatcher matcher = new RealTimeRegexpMatcher(regexQuery, fst, dest);
+    matcher.regexMatchOnFST();
+  }
+
+  // Matches "input" string with _regexQuery Automaton.
+  public boolean match(String input) {
+    CharacterRunAutomaton characterRunAutomaton = new CharacterRunAutomaton(_automaton);
+    return characterRunAutomaton.run(input);
+  }
+
+  /**
+   * This function runs matching on automaton built from regexQuery and the FST.
+   * FST stores key (string) to a value (Long). Both are state machines and state transition is based on
+   * a input character.
+   *
+   * This algorithm starts with Queue containing (Automaton Start Node, FST Start Node).
+   * Each step an entry is popped from the queue:
+   *    1) if the automaton state is accept and the FST Node is final (i.e. end node) then the value stored for that FST
+   *       is added to the set of result.
+   *    2) Else next set of transitions on automaton are gathered and for each transition target node for that character
+   *       is figured out in FST Node, resulting pair of (automaton state, fst node) are added to the queue.
+   *    3) This process is bound to complete since we are making progression on the FST (which is a DAG) towards final
+   *       nodes.
+   */
+  public void regexMatchOnFST() {
+    final List<Path> queue = new ArrayList<>();
+
+    if (_automaton.getNumberOfStates() == 0) {
+      return;
+    }
+
+    // Automaton start state and FST start node is added to the queue.
+    queue.add(new Path(_automaton.getInitialState(), _fst.getStartState(), null, new ArrayList<>()));
+
+    Set<State> acceptStates = _automaton.getAcceptStates();
+    while (queue.size() != 0) {
+      final Path path = queue.remove(queue.size() - 1);
```

Comment:
When you change this to `ArrayDeque` you can just call `queue.removeLast()` instead. This is exactly the sort of thing `Deque` is for.